packages/babel-helpers/src/helpers/applyDecs.js

```diff
@@ -526,27 +605,43 @@ function applyClassDecs(ret, targetClass, metadataMap, classDecs) {
   if (classDecs.length > 0) {
     var initializers = [];
     var newClass = targetClass;
-
     var name = targetClass.name;
-    var ctx = Object.assign(
-      {
-        kind: "class",
-        name: name,
-        addInitializer: createAddInitializerMethod(initializers),
-      },
-      createMetadataMethodsForProperty(metadataMap, 0 /* CONSTRUCTOR */, name)
-    );
 
     for (var i = classDecs.length - 1; i >= 0; i--) {
-      newClass = classDecs[i](newClass, ctx) || newClass;
+      var decorationState = { finished: false };
+
+      var ctx = Object.assign(
+        {
+          kind: "class",
+          name: name,
+          addInitializer: createAddInitializerMethod(
+            initializers,
+            decorationState
+          ),
+        },
+        createMetadataMethodsForProperty(
+          metadataMap,
+          0 /* CONSTRUCTOR */,
+          name,
+          decorationState
+        )
+      );
+
+      var nextNewClass = classDecs[i](newClass, ctx);
+      if (nextNewClass !== undefined) {
+        assertValidReturnValue(10 /* CLASS */, nextNewClass);
+        newClass = nextNewClass;
+      }
+
+      decorationState.finished = true;
```

Comment:
Q: Should we also mark state as finished if applying decorator results to an abrupt completion?

I think we should, so the leaked context method will not be allowed to call again in a catch clause outside the class, but of course one can still use try-catch within the decorator.